include/boost/graph/isomorphism.hpp

```diff
@@ -165,18 +163,26 @@ namespace detail
             BGL_FORALL_VERTICES_T(v, G1, Graph1)
             f[v] = graph_traits< Graph2 >::null_vertex();
 
+            std::size_t max_invariant = 0;
             {
                 std::vector< invar1_value > invar1_array;
+                invar1_array.reserve(num_vertices(G1));
                 BGL_FORALL_VERTICES_T(v, G1, Graph1)
-                invar1_array.push_back(invariant1(v));
+                  invar1_array.push_back(invariant1(v));
                 sort(invar1_array);
 
                 std::vector< invar2_value > invar2_array;
+                invar2_array.reserve(num_vertices(G2));
                 BGL_FORALL_VERTICES_T(v, G2, Graph2)
-                invar2_array.push_back(invariant2(v));
+                  invar2_array.push_back(invariant2(v));
                 sort(invar2_array);
                 if (!equal(invar1_array, invar2_array))
                     return false;
+
+                // Empty graphs case is covered before test_isomorphism is
+                // called, so the invar?_arrays cannot be empty, so back() is
+                // safe. Also the two invariant arrays are equal:
+                max_invariant = invar1_array.back() + 1;
```

Comment:
What if `invar1_array.back() == std::numeric_limits<std::size_t>::max()`?

It will happen eventually.

I had a quick look through the whole code for isomorphism but it's not obvious to me exactly what the simplest way of dealing with this is.

One option is that we only support values less than `std::numeric_limits<std::size_t>::max()` in the array and guard against it defensively using `assert`. I think this is reasonable.

Reply:
But I'm totally open to other suggestions.

Reply:
Well, there is the documentation change from

```diff
- The values returned by these two functions must be in the range [0, <tt>vertex_max_invariant</tt>).
+ The values returned by these two functions must have a low upper bound. 
```

The reason that's there is not the possibility of overflow, but rather the [invariant multiplicity counting](https://github.com/jargonzombies/graph/blob/ec286110269fc297bd3b3b6b687379b7d0afd8d3/include/boost/graph/isomorphism.hpp#L192), which allocates a zeroed vector of length equal to `max_invariant`, which is what I found when I encountered an out-of-memory error in trying a colored graph isomorphism with hashed invariants, prompting #203.

So to some degree I think adding an `assert` papers over a more fundamental issue that might want addressing.

The fix I would suggest is a [multiplicity counting using `unordered_map`](https://github.com/boostorg/graph/pull/204/files#diff-23655e4e06fd65a235692ad0debfc5c08c59dc6ba4cd4b27424c66df437b0bd8R177) instead, which eliminates `max_invariant` entirely, and allows consumers to use the full `size_t` range in their invariants without allocating excessive memory. Assuming this PR and #206 (which introduces `unordered_map` ) are accepted, I would have introduced another PR adding the remaining bits of #204, which, besides the counting, are just hashable concept checks on the invariant function result type and documentation changes.

I think there are two options:

- We add the multiplicity counting algorithm here.
- We add the assert and expand the documentation here, mentioning the allocation to explain the low bound documentation addition. If the follow-up PR with multiplicity counting gets accepted, that gets removed again.

I'm going to go with the latter for now. Let me know if you disagree.

Reply:
I'm inclined to trust your judgement here and go with whatever is the most efficient way for you to get it done. Sorry about wasting time with splitting the PR into multiple pieces, but at the time I felt like I wasn't going to be able to process it all at once.

Reply:
I do think it's better split up the way it is, so no worries. The original PR did do a lot of things at once.